### ISSUE TITLE
Conformance: sigstore-python's conformance runner should support `--trusted-root`

### ISSUE THREAD

**Issue by woodruffw** [sigstore/sigstore-python#821]
See https://github.com/sigstore/sigstore-conformance/pull/101.

**Comment by woodruffw** [comment]
Partial dupe of #779, closing that one in favor of this.

**Comment by woodruffw** [comment]
> * deprecate flags mentioned above to make it clear a complete trusted root should be provided instead (there might be some exceptions to this but I think that's a good general rule)

Yeah, I _think_ we can get all of them with this -- we could deprecate with 2.1.x and plan to fully remove with 3.x, which would make me very happy 🙂

**Comment by jku** [comment]
Some notes after a bit of thinking:
* We have essentially three cases (only first one is implemented today)
  * default: trustroot comes from TUF
  * offline: trustroot comes from TUF target cache
  * trustroot comes from user via `--trusted-root`
* Current TrustedRoot class handles two separate things: the TUF client operations and easy access to trustroot contents
  * We should separate them as the TUF operations are only needed in the default case above but the latter is needed in all three cases

**Comment by woodruffw** [comment]
> * We should separate them as the TUF operations are only needed in the default case above but the latter is needed in all three cases

Makes sense to me!

**Comment by jku** [comment]
Okay I have the internal refactor done... but I'm not sure how to correctly add the --trusted-root flag: 
* I should choose the fulcio_url and rekor_url values based on data from trusted root
* I can see the correct values in the file...
* but it looks like I could have multiple values of each -- is someone checking that each object in `certificateAuthorities` has the same uri and and that each object in `tlogs` has the same baseURL?
* Can I just assume the first object has the value I want?

@woodruffw I'm guessing you'll have a hunch on this

**Comment by jku** [comment]
And another question:
* for signing we need to know the OAuth issuer URL. Where does that come from?

**Comment by jku** [comment]
> I should choose the rekor_url value based on data from trusted root: can I assume all values in tlogs array have the same baseURL? Maybe RekorClient should handle url per key?

I think I'll look into modifying the Keyring abstraction so that it keeps track of url-key pairs instead of just keys. Alternative ideas are welcome.

**Comment by woodruffw** [comment]
> Maybe RekorClient should handle url per key?

This sounds right to me -- my read of the `TrustedRoot` definition is that the base URLs are not guaranteed to be equivalent between different log instances (and maybe even can't be, since the point is to have multiple distinct logs?)

> same url question for CAs...

Yeah, this is confusing/ambiguous: my intuition there is that the `TrustedRoot` needs some way to signal the "CSR" CA, i.e. the one that's actually talked to for signing purposes. But that doesn't seem to exist yet, so maybe the right logic here (for now) is to fail unless there's exactly one CA listed *and* that CA has a URI?

> oauth issuer url is not documented in the trusted_root -- maybe it should be

This would be the Dex instance URL, right? If so, yes, agreed.

**Comment by woodruffw** [comment]
> I think I'll look into modifying the Keyring abstraction so that it keeps track of url-key pairs instead of just keys. Alternative ideas are welcome.

> Alternatively I might stop here and make a refactor PR out of my current changes... So then we can do a spearate refactoring for keyring/rekorClient and then add the UI option

That keyring refactor SGTM, but 👍 on creating an initial refactor PR first -- smaller changesets will be easier to test 🙂

**Comment by jku** [comment]
I think we agree on everything here. 
* I'll make a refactoring PR #844
* I'll make an issue about refactoring KeyRing/RekorClient
* this issue (adding --trusted-root) can be fixed after that

Since that all might take a while we may want to use xfail in the workflow in the mean time so we can start using sigstore-conformance 0.0.9.
